File: Olympiad/COCI/official/2007/contest4_solutions/ispiti.cpp

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 4 - Task ISPITI
  Programming language C++
*/

#include <algorithm>
#include <cstdio>
#include <cstring>

using namespace std;

int n, m;
struct ucenik {
   int a, b;
   int sorted_index; // indeks ucenika u sortiranom nizu ucenika
   int index;  // indeks ucenika u inputu
} U[200000];

bool pair_cmp( const ucenik &X, const ucenik &Y ) { // sortiranje po drugom, pa prvom broju
   if( X.b != Y.b ) return X.b < Y.b;
   return X.a < Y.a;
}

bool index_cmp( const ucenik &X, const ucenik &Y ) { // sortiranje po indeksu ucenika
   return X.index < Y.index;
}

struct event { // jedan dogadjaj iz inputa: type = 1 za dolazak ucenika, a 2 za upit
   int who, type;
} E[200000];

int tree[1<<19];

inline int bolji( int uc1, int uc2 ) { // usporedjuje po prvom broju ucenike, i pazi na prazna mjesta u stablu
   if( uc1 == -1 ) return uc2;
   if( uc2 == -1 ) return uc1;
   if( U[uc1].a < U[uc2].a ) return uc2;
   return uc1;
}

void update( int i, int lo, int hi, int uc ) { // dodaj ucenika uc u tournament tree
   if( lo > U[uc].sorted_index || hi < U[uc].sorted_index ) return;

   if( lo == hi ) { // ako smo na listu stabla, zapisi uc
      tree[i] = uc;
   } else {
      int mid = (lo+hi)/2;
      update( 2*i, lo, mid, uc ); // odi lijevo
      update( 2*i+1, mid+1, hi, uc ); // odi desno

      tree[i] = bolji( tree[2*i], tree[2*i+1] ); // u trenutni zapisi "bolji" od dva podstabla
   }
}

// i u query i u update proceduri, varijabla i pamti trenutnu poziciju u stablu, a lo i hi pamte u kojem podintervalu trazimo

int query( int i, int lo, int hi, int uc ) { // vrati prvi desno koji ima prvi broj veci od ucenika uc
   if( hi <= U[uc].sorted_index ) return -1; // ako se uc nalazi ljevo od podintervala [lo, hi) vrati -1
   if( tree[i] == -1 || U[tree[i]].a < U[uc].a ) return -1; // ako u podintervalu nema ucenika ili je najveci prvi broj manji od onoga u uc, vrati -1

   if( lo == hi ) return tree[i];

   int mid = (lo+hi)/2; // novi podintervali su [lo, mid) i [mid, hi)

   int t = query( 2*i, lo, mid, uc );
   if( t != -1 ) return t; // ako u lijevom podintervalu ima netko veci od njega, vrati rezultat, nema potrebe da idemo u desni
   return query( 2*i+1, mid+1, hi, uc ); // vrati prvi veci u desnom podintervalu

}

int main( void ) {
   scanf( "%d", &n );
   for( int i = 0; i < n; ++i ) { // prvo sve ucitamo, onda rjesavamo problem
      char c;
      scanf( " %c", &c );
      if( c == 'D' ) {
         scanf( "%d%d", &U[m].a, &U[m].b );
         U[m].index = m;
         E[i].who = m++;
         E[i].type = 1;
      } else {
         scanf( "%d", &E[i].who ); --E[i].who;
         E[i].type = 2;
      }
   }

   sort( U, U+m, pair_cmp ); // sortiramo ucenike po drugom broju, a po prvom u slucaju izjednacenja
   for( int i = 0; i < m; ++i ) U[i].sorted_index = i; // sazimanje
   sort( U, U+m, index_cmp ); // vracamo ih kako su bili, da mozemo lako doci do i-tog ucenika

   memset( tree, -1, sizeof tree );
   for( int i = 0; i < n; ++i ) {
      if( E[i].type == 1 ) {
         update( 1, 0, m-1, E[i].who ); // doselio se ucenik E[i].who
      } else {
         int x = query( 1, 0, m-1, E[i].who ); // nadji prvi desno u tree-u od ucenika E[i].who, koji ima veci prvi broj
         if( x == -1 ) printf( "NE\n" );
         else printf( "%d\n", x+1 );
      }
   }

   return 0;
}

File: Olympiad/COCI/official/2007/contest4_solutions/ispiti.pas

(*
  Croatian Open Competition in Informatics 2006/2007
  Contest 4 - Task ISPITI
  Programming language Pascal
*)

program ispiti;

const maxn = 200000;
      maxt = 524288;

type event = record    (* jedan dogadjaj iz ulaza: type = 1 za dolazak ucenika, a 2 za upit *)
    who, tip: longint;
end;

type ucenik = record
    a, b: longint;
    index: longint; (* indeks ucenika u sortiranom nizu ucenika *)
    sorted_index: longint; (* indeks ucenika u ulazu *)
end;

var
    m, n, i, j: longint;
    tree: array[0..maxt] of longint;
    U: array[0..maxn] of ucenik;
    E: array[0..maxn] of event;
    c: char;

function pair_cmp( x, y: ucenik ):boolean; begin (* sortiranje po drugom, pa prvom broju *)
    if x.b <> y.b then pair_cmp := (x.b < y.b)
    else pair_cmp := (x.a < y.a);
end;

function index_cmp( x, y: ucenik ):boolean; begin (* sortiranje po indeksu ucenika *)
    index_cmp := (x.index < y.index);
end;

function master_cmp( x,y: ucenik; k: longint ):boolean; begin
    if k = 0 then master_cmp := pair_cmp( x, y )
    else master_cmp := index_cmp( x, y );
end;

function bolji( uc1, uc2: longint ):longint; begin  (* usporedjuje po prvom broju ucenike, i pazi na prazna mjesta u stablu *)
    if uc1 = -1 then begin bolji := uc2; exit; end;
    if uc2 = -1 then begin bolji := uc1; exit; end;
    if U[uc1].a < U[uc2].a then bolji := uc2
    else bolji := uc1;
end;

procedure update( i, lo, hi, uc: longint ); (* dodaj ucenika uc u tournament tree *)
var mid: longint;
begin
    if (lo > U[uc].sorted_index) or (hi < U[uc].sorted_index) then exit;

    if lo = hi then tree[i] := uc (* ako smo na listu stabla, zapisi uc *)
    else begin
        mid := (lo+hi) div 2;

        update( 2*i, lo, mid, uc ); (* odi lijevo *)
        update( 2*i+1, mid+1, hi, uc ); (* odi desno *)

        tree[i] := bolji( tree[2*i], tree[2*i+1] ); (* u trenutni zapisi "bolji" od dva podstabla *)
    end;
end;

(* i u query i u update proceduri, varijabla i pamti trenutnu poziciju u stablu, a lo i hi pamte u kojem podintervalu trazimo *)

function query( i, lo, hi, uc: longint ): longint; (* vrati prvi desno koji ima prvi broj veci od ucenika uc *)
var t, mid: longint;
begin
    if hi <= U[uc].sorted_index then
        begin query := -1; exit; end; (* ako se uc nalazi ljevo od podintervala [lo, hi) vrati -1 *)

    if (tree[i] = -1) or (U[tree[i]].a < U[uc].a) then
        begin; query := -1; exit; end; (* ako u podintervalu nema ucenika ili je najveci prvi broj manji od onoga u uc, vrati -1 *)

    if lo = hi then begin query := tree[i]; exit; end;

    mid := (lo+hi) div 2; (* novi podintervali su [lo, mid) i [mid, hi) *)

    t := query( 2*i, lo, mid, uc );
    if t <> -1 then query := t (* ako u lijevom podintervalu ima netko veci od njega, vrati rezultat, nema potrebe da idemo u desni *)
    else query := query( 2*i+1, mid+1, hi, uc ); (* vrati prvi veci u desnom podintervalu *)
end;

procedure load; begin
    readln( n );
    for i := 0 to n-1 do begin (* prvo sve ucitamo, onda rjesavamo problem *)
        read( c );
        if c = 'D' then begin
            readln( U[m].a, U[m].b );
            U[m].index := m;
            E[i].who := m;
            inc( m );
            E[i].tip := 1;
        end else begin
            readln( E[i].who ); dec( E[i].who );
            E[i].tip := 2;
        end;
    end;
end;

procedure swap( var a, b: ucenik );
var tmp: ucenik;
begin
    tmp := a;
    a := b;
    b := tmp;
end;

procedure sortiraj( l, r, k: longint ); (* qsort *)
var mid: ucenik;
    x,y: longint;
begin
    if r-l < 2 then exit;
    if r-l = 2 then begin
        if master_cmp( U[l+1], U[l], k ) then swap( U[l], U[l+1] );
        exit;
    end;

    mid := U[(l+r) div 2];
    y := l;

    for x := l to r-1 do  (* sve strogo manje stavi na pocetak *)
        if( master_cmp( U[x], mid, k ) ) then begin
            swap( U[x], U[y] );
            inc( y );
        end;

    sortiraj( l, y, k ); (* sortiraj sve strogo manje *)

    for x := y to r-1 do (* sve jednake stavi poslije ovih manjih *)
        if( ( (k = 0) and (U[x].b = mid.b) and (U[x].a = mid.a) ) or ( (k = 1) and (U[x].index = mid.index) ) ) then begin
            swap( U[x], U[y] );
            inc( y );
        end;

    sortiraj( y, r, k ); (* sortiraj sve strogo vece *)
end;

procedure sazmi; begin   (* sazimanje po broju b *)
    sortiraj( 0, m, 0 );
    for i := 0 to m-1 do U[i].sorted_index := i;
    sortiraj( 0, m, 1 );
end;

procedure solve; begin    (* rjesavanje upita *)
    for i := 0 to maxt-1 do tree[i] := -1;
    for i := 0 to n-1 do begin
        if E[i].tip = 1 then update( 1, 0, m-1, E[i].who )
        else begin
            j := query( 1, 0, m-1, E[i].who );
            if j = -1 then writeln( 'NE' )
            else writeln( j+1 );
        end;
    end;
end;

begin
    load;
    sazmi;
    solve;
end.
